lib/spack/spack/binary_distribution.py

```diff
@@ -133,6 +131,349 @@ def __init__(self, errors):
+        """Retrieve the archive file and return the local archive file path"""
+        if not self.spec_dict:
+            raise BuildcacheEntryError(
+                "Cannot fetch archive without first fetching valid metadata"
+            )
+
+        if not allow_unsigned:
+            if not self.has_signed:
+                raise BuildcacheEntryError(
+                    "Buildcache entry created with allow_unsigned=False, but no signed "
+                    "spec file exists on mirror"
+                )
+            self.verified = try_verify(self.get_local_spec_path())
+            if not self.verified:
+                raise NoVerifyException(
+                    f"Signature on {self.remote_spec_url} could not be verified"
+                )
+
+        self.archive_stage = spack.stage.Stage(self.remote_archive_url)
+
+        # Fetch the archive file, or else cleanup and exit early
+        try:
+            self.archive_stage.create()
+            self.archive_stage.fetch()
+        except spack.error.FetchError as e:
+            self.destroy()
+            raise BuildcacheEntryError("Unable to fetch archive") from e
+
+        self.local_archive_path = self.archive_stage.save_filename
+
+        return self.local_archive_path
+
+    def get_relative_spec_components(self, spec: spack.spec.Spec, ext: str) -> List[str]:
+        spec_formatted = spec.format_path("{name}-{version}-{hash}")
+        return [f"v{self.layout_version}", "specs", f"{spec_formatted}{ext}"]
+
+    def get_relative_tarball_components(self, algorithm: str, checksum: str) -> List[str]:
+        return ["blobs", algorithm, checksum[:2], checksum]
+
+    def get_relative_keys_components(self) -> List[str]:
+        return [f"v{self.layout_version}", "keys", "_pgp"]
+
+    def get_relative_specs_components(self) -> List[str]:
+        return [f"v{self.layout_version}", "specs"]
+
+    def get_relative_blobs_components(self) -> List[str]:
+        return ["blobs"]
+
+    def get_remote_spec_url(self):
+        return self.remote_spec_url
+
+    def get_remote_archive_url(self):
+        return self.remote_archive_url
+
+    def get_local_spec_path(self):
+        return self.local_specfile_path
+
+    def get_local_archive_path(self):
+        return self.local_archive_path
+
+    def get_signature_required(self):
+        return self.signature_required
+
+    def get_archive_checksum_algorithm(self):
+        return self.remote_archive_checksum_algorithm
+
+    def get_archive_checksum_hash(self):
+        return self.remote_archive_checksum_hash
+
+    def destroy(self):
+        if self.spec_stage:
+            self.spec_stage.destroy()
+            self.spec_stage = None
+        if self.archive_stage:
+            self.archive_stage.destroy()
+            self.archive_stage = None
+
+    def is_verified(self):
+        return self.verified
+
+    def get_spec_dict(self) -> Dict:
+        return self.spec_dict
+
+
+def create_urlbuildcacheentry(
```

Comment:
underscores between words

Reply:
You think so here?  `urlbuildcacheentry` here comes from the name of the class, putting underscores in between seems to make that less clear to me.  But I'm not going to argue it too much if you insist.

Reply:
I agree this is not consistent style